app/services/saml_request_validator.rb

```diff
@@ -25,8 +25,10 @@ def extra_analytics_attributes
     }
   end
 
+  # This check relies on the fact that problematic SPs are returned as NullServiceProvider objects.
+  # It should be disentangled and SP errors should be validated explicitly.
   def authorized_service_provider
-    return if service_provider.active? # live? instead when dashboard approvals matter.
+    return if service_provider.active? || !service_provider.is_a?(NullServiceProvider)
```

Comment:
the `NullServiceProvider` object is one that we control....we can just make its `#active?` method return false right?

Reply:
Yeah, this is where things gets weird. We utilize the fact that the NSP is `active = false` as a proxy for any problem with the issuer, whether that be the null byte issue Mitchell fixed a couple of weeks, issuer blank, or issuer not found. All of that is caught because through that `service_provider.active?` check. So now, we don't have a way to differentiate between an invalid SP (caught by the fact that it defaults to `active = false`) and a valid SP that is not active.

I spent a lot of time trying to find a reasonable refactor for this issue without really reorganizing this code I don't feel comfortable with yet. And made all the more complicated by the fact that the same issue exists in both OIDC and SAML but their implementations are so different that without a massive refactor they would have to be fixed more or less separately.

Thus this NSP check in both, while admittedly ugly, allows us to maintain current functionality without a pretty significant refactor.

Reply:
Can you file a follow-up ticket to clean up this NullServiceProvider check? The more I think about this the more I feel like there has to be another method we can check agains

Reply:
Will do!